### PR TITLE
Foreman operatingsystem

### PR DESCRIPTION
**Pull request by mdellweg**
This adds yet another module. Hurray!
This time for Operating systems.

### CODE REVIEW

**Comment by mdellweg** [comment]
This PR relies on:
https://github.com/SatelliteQE/nailgun/pull/448
and it's update functionality is blocked by:
http://projects.theforeman.org/issues/21169

**Comment by Fobhep** [comment]
@evgeni @akofink just leaving a friendly ping here ;)

File: module_utils/ansible_nailgun_cement.py [hunk 1]
```diff
@@ -19,7 +19,9 @@
     Repository,
     RepositorySet,
     TemplateKind,
-    AbstractComputeResource
+    AbstractComputeResource,
+    OSDefaultTemplate,
+    _check_for_value,
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply module_utils/ansible_nailgun_cement.py]: Oh, yes, this is a left over from me, trying to bridge the necessary nailgun gap in cement. Just before i decided to do it there: https://github.com/SatelliteQE/nailgun/pull/448

File: README.md [hunk 2]
```diff
@@ -61,6 +61,7 @@ This is a list of modules currently in the repository (please add to the list if
 #### Entity Modules
 
  * foreman_global_parameter: create and maintain global parameters
+ * foreman_operating_system: create and maintain operating systems
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply README.md]: Will be added.

**Comment by akofink** [comment]
I don't have strong feelings either way. You make a good point about the different params available during create and update.

File: test/demo/operating_system.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+---
+- hosts: localhost
+  tasks:
+  - name: "Create Debian operating system"
+    foreman_os_default_template:
```
Comment (akofink) [inline test/demo/operating_system.yml]: `foreman_operating_system`

File: modules/foreman_os_default_template.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    required: true
+    choices:
+      - present
+      - present_with_defaults
+      - absent
+'''
+
+EXAMPLES = '''
+- name: "Create an Association"
+  foreman_operating_system:
```
Comment (akofink) [inline modules/foreman_os_default_template.py]: `foreman_os_default_template`

File: modules/foreman_os_default_template.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    operatingsystem: "CoolOS"
+    template_kind: "finish"
+    provisioning_template: "CoolOS finish"
+    state: present
+
+- name: "Delete an Association"
+  foreman_operating_system:
```
Comment (akofink) [inline modules/foreman_os_default_template.py]: `foreman_os_default_template`

File: modules/foreman_os_default_template.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    state = entity_dict.pop('state')
+
+    try:
+        create_server(server_url, (username, password), verify_ssl)
+    except Exception as e:
+        module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
+    ping_server(module)
+    entity_dict['operatingsystem'] = find_entities_by_name(
+        OperatingSystem, [entity_dict['operatingsystem']], module)[0]
```
Comment (akofink) [inline modules/foreman_os_default_template.py]: I don't think `OperatingSystem` names are unique. We should really be doing a title search using '~':

```python
OperatingSystem.search(set(), {'search': 'title~"{}"'.format(entity_dict['operatingsystem'])})
```

This allows a distinction between, say, Debian 9.1 and Debian 9.2.

**Comment by akofink** [comment]
The changes worked well during my tests. There's just a few typos and one update to searching for operating systems, and it'll be good to go.

**Comment by akofink** [comment]
Whoops - should have waited on Nailgun PR. :see_no_evil:

**Comment by mdellweg** [comment]
To bad Github!
@akofink Yes probably waiting would not have been bad. But a bird told me you have write access there...?!? Also, at least it is approved. 🙉

**Comment by akofink** [comment]
Sadly, I do not have Nailgun permissions.

**Comment by mdellweg** [comment]
You could add a label for do not merge.

**Comment by mdellweg** [comment]
Or as this will probably be a recurring situation, a `wait for nailgun` label.
Or we just ship our own version of Nailgun... 🤦‍

**Pull request merged**